lib/openssh/sshd.go

```diff
@@ -0,0 +1,254 @@
+// teleport specific configuration
+func (s *SSHD) UpdateConfig(u SSHDConfigUpdate, restart bool) error {
+	configUpdate, err := fmtSSHDConfigUpdate(u)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := writeTempAndRename(filepath.Join(u.DataDir, sshdConfigFile), s.sshd.checkConfig, []byte(configUpdate)); err != nil {
+		return trace.Wrap(err)
+	}
+
+	sshdConfigInclude := sshdConfigInclude(u.DataDir)
+
+	needsUpdate, err := checkSSHDConfigAlreadyUpdated(u.SSHDConfigPath, sshdConfigInclude)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if needsUpdate {
+		if err := prependToSSHDConfig(u.SSHDConfigPath, sshdConfigInclude); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if restart {
+		if err := s.sshd.restart(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// WriteKeys writes the OpenSSH keys and CA from the Identity and the
+// OpenSSH CA to disk for the OpenSSH daemon to use
+func WriteKeys(keysdir string, id *auth.Identity, cas []types.CertAuthority) error {
+	if err := os.MkdirAll(keysdir, 0o700); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := writeTempAndRename(filepath.Join(keysdir, TeleportKey), func(_ string) error { return nil }, id.KeyBytes); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.Chmod(filepath.Join(keysdir, TeleportKey), 0o600); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := writeTempAndRename(filepath.Join(keysdir, TeleportCert), func(_ string) error { return nil }, id.CertBytes); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	var caKeyBytes []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			pubKey := append(bytes.TrimSpace(key.PublicKey), byte('\n'))
+			caKeyBytes = append(caKeyBytes, pubKey...)
+		}
+	}
+
+	if err := writeTempAndRename(filepath.Join(keysdir, TeleportOpenSSHCA), func(_ string) error { return nil }, caKeyBytes); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+type sshdBackend struct {
+	restartCmd string
+}
+
+var _ sshdBackendOperations = &sshdBackend{}
+
+func (*sshdBackend) checkConfig(path string) error {
+	cmd := exec.Command(sshdBinary, "-t", "-f", path)
+	if err := cmd.Run(); err != nil {
+		output, outErr := cmd.CombinedOutput()
+		if err != nil {
+			return trace.Wrap(trace.NewAggregate(err, outErr), "invalid sshd config file, failed to get `sshd -t` output")
+		}
+		return trace.Wrap(err, "invalid sshd config file %q, not writing", string(output))
+	}
+	return nil
+}
+
+func (b *sshdBackend) restart() error {
+	cmd := exec.Command(sshdBinary, "-t")
+	if err := cmd.Run(); err != nil {
+		output, outErr := cmd.CombinedOutput()
+		if outErr != nil {
+			return trace.Wrap(trace.NewAggregate(err, outErr), "invalid sshd config file, failed to get `sshd -t` output")
+		}
+		return trace.Wrap(err, "invalid sshd config file %q, not writing", string(output))
+	}
+
+	cmd = exec.Command("/bin/sh", "-c", b.restartCmd)
```

Comment:
Why do we need to call this via shell?

Reply:
Ah I didn't realize this command is being passed from the config and is a single string. I guess that's fine as-is.